pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,640 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cronjob
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/robfig/cron"
+
+	batchv1 "k8s.io/api/batch/v1"
+	batchv1beta1 "k8s.io/api/batch/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	batchv1informers "k8s.io/client-go/informers/batch/v1"
+	batchv1beta1informers "k8s.io/client-go/informers/batch/v1beta1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	covev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	batchv1listers "k8s.io/client-go/listers/batch/v1"
+	batchv1beta1listers "k8s.io/client-go/listers/batch/v1beta1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/component-base/metrics/prometheus/ratelimiter"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/controller"
+)
+
+var (
+	nextScheduleDelta = 100 * time.Millisecond
+)
+
+// Refactored Cronjob controller that uses DelayingQueue and informers
+
+// ControllerV2 is a controller for CronJobs.
+type ControllerV2 struct {
+	queue    workqueue.DelayingInterface
+	recorder record.EventRecorder
+
+	jobControl     jobControlInterface
+	cronJobControl cjControlInterface
+
+	jobLister     batchv1listers.JobLister
+	cronJobLister batchv1beta1listers.CronJobLister
+
+	jobListerSynced     cache.InformerSynced
+	cronJobListerSynced cache.InformerSynced
+
+	// now is a function that returns current time, done to facilitate unit tests
+	now func() time.Time
+}
+
+// NewControllerV2 creates and initializes a new Controller.
+func NewControllerV2(jobInformer batchv1informers.JobInformer, cronJobsInformer batchv1beta1informers.CronJobInformer, kubeClient clientset.Interface) (*ControllerV2, error) {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
```

Comment:
we're migrating to StartStructuredLogging(0) now

Reply:
updated as suggested